sylt-parser/src/statement.rs

```diff
@@ -427,71 +429,101 @@ pub fn statement<'t>(ctx: Context<'t>) -> ParseResult<'t, Statement> {
-                            };
-                            (ctx_, ty)
-                        };
-                        ctx = ctx_;
-                        variants.insert(variant, ty);
-                        ctx = ctx.skip_if(T::Comma);
-                        ctx = ctx.skip_if(T::Newline);
+            let (ctx, variables) =
+                parse_beg_end_comma_sep!(ctx, T::LeftParen, T::RightParen, &item)?;
+
+            // Parse variants: `A(..), B(..)`
+            let (ctx, items) = {
+                fn sep<'t>(ctx: Context<'t>) -> ParseResult<'t, ()> {
+                    let ctx = skip_while!(ctx, T::Newline);
+                    let ctx = ctx.skip_if(T::Comma);
+                    Ok((ctx, ()))
+                }
+                fn end<'t>(ctx: Context<'t>) -> ParseResult<'t, bool> {
+                    let ctx = skip_while!(ctx, T::Newline);
+                    match ctx.token() {
+                        // Done with variants.
+                        T::End => Ok((ctx.skip(1), true)),
+                        _ => Ok((ctx, false)),
                     }
+                }
 
-                    _ => {
+                fn item<'t>(ctx: Context<'t>) -> ParseResult<'t, (Identifier, Type)> {
+                    let ctx = skip_while!(ctx, T::Newline);
+                    let (ctx, variant) = match ctx.eat() {
+                        (T::Identifier(variant), span, ctx) => {
+                            (ctx, Identifier::new(span, variant.clone()))
+                        }
+                        _ => raise_syntax_error!(
+                            ctx,
+                            "An enum variant has to start with a valid identifier"
+                        ),
+                    };
+                    if !is_capitalized(&variant.name) {
                         raise_syntax_error!(
                             ctx,
-                            "Expected variant name or 'end' in enum statement"
+                            "Enum variants have to start with a capital letter"
                         );
                     }
+                    let (ctx, ty) = if matches!(ctx.token(), T::End | T::Comma | T::Newline) {
+                        let ty = Type {
+                            span: variant.span,
+                            kind: TypeKind::Resolved(RuntimeType::Nil),
+                        };
+                        (ctx, ty)
+                    } else {
+                        let ctx = ctx.skip_if(T::Colon);
+                        let (ctx, ty) = parse_type(ctx)?;
+                        if !matches!(ctx.token(), T::Comma | T::End | T::Newline) {
+                            raise_syntax_error!(ctx, "Expected a delimiter ','");
+                        };
+                        (ctx, ty)
+                    };
+                    Ok((ctx.skip_if(T::Comma), (variant, ty)))
                 }
+                parse_sep_end_by(ctx, &sep, &end, &item)?
```

Comment:
i _think_ function pointers are automatically borrowed

Reply:
I had to add it. Tried to remove it.

Reply:
I think impl solved this too :o: